File: src/content/docs/topics/cs/leetcode-150/graphs/127-word-ladder.py
def ladder_length(beginWord, endWord, wordList):
    pass  # TODO: implement

def _run_tests():
    assert ladder_length('hit', 'cog', ['hot', 'dot', 'dog', 'lot', 'log', 'cog']) == 5
    assert ladder_length('hit', 'cot', ['hot', 'dot', 'dog', 'lot', 'log', 'cog']) == 0
    assert ladder_length('hot', 'dot', ['dot', 'lot']) == 2
    assert ladder_length('hit', 'cog', ['hot', 'dot', 'dog', 'lot', 'log']) == 0
    assert ladder_length('a', 'c', ['a', 'b', 'c']) == 2
    # --- large-input timing ---
    import time as _t
    _t0 = _t.perf_counter()
    ladder_length('aaaa', 'zzzz', [chr(65 + i % 26) * 4 for i in range(200)])
    _ms = (_t.perf_counter() - _t0) * 1000
    print(f'perf word-ladder 200-word list: {_ms:.1f}ms')
    print('all tests pass')

if __name__ == '__main__':
    _run_tests()

File: src/content/docs/topics/cs/coding-problems/graphs/127-word-ladder.swift
// LEETCODE_TYPE: Solution
func expectEqual<T: Equatable>(
    _ actual: T,
    _ expected: T,
    _ message: String = "",
    file: StaticString = #fileID,
    line: UInt = #line
) {
    guard actual == expected else {
        let detail = message.isEmpty ? "values differ" : message
        fatalError("\(file):\(line): \(detail). Expected \(expected), got \(actual)")
    }
}

func expectTrue(
    _ condition: @autoclosure () -> Bool,
    _ message: String = "expected true",
    file: StaticString = #fileID,
    line: UInt = #line
) {
    guard condition() else {
        fatalError("\(file):\(line): \(message)")
    }
}

func reportSuccess() {
    print("All Swift tests passed")
}
final class Solution {
    func ladderLength(_ beginWord: String, _ endWord: String, _ wordList: [String]) -> Int {
        // TODO: Implement
        fatalError("TODO: Implement")
    }
}

func runTests() {
    // TEST_VECTORS_BEGIN sha256:4b2414e123fbe73a4543101ac0e3fd2011b92f9ad1e98f6beb78be0aa8bd9f20
    expectEqual(Solution().ladderLength("hit", "cog", ["hot", "dot", "dog", "lot", "log", "cog"]), 5, "classic-ladder")
    expectEqual(Solution().ladderLength("hit", "cog", ["hot", "dot", "dog", "lot", "log"]), 0, "missing-end-word")
    expectEqual(Solution().ladderLength("a", "c", ["a", "b", "c"]), 2, "one-change")
    // EXCLUDED_VECTOR same-endpoints: ["hit","hit",["hit"]] | The problem contract requires distinct begin and end words.
    // TEST_VECTORS_END
    reportSuccess()
}

runTests()
